identity_credential/src/validator/domain_linkage_validator.rs

```diff
@@ -0,0 +1,265 @@
+        let expected_types = HashSet::from([
+          Credential::<T>::base_type(),
+          DomainLinkageConfiguration::domain_linkage_type(),
+        ]);
+        let types = HashSet::from([type_1.as_str(), type_2.as_str()]);
+        if !types.eq(&expected_types) {
+          Err(DomainLinkageVerificationError {
+            cause: DomainLinkageVerificationErrorCause::InvalidTypeProperty,
+            source: None,
+          })?;
+        }
+      }
+      OneOrMany::One(_) => {
+        Err(DomainLinkageVerificationError {
+          cause: DomainLinkageVerificationErrorCause::InvalidTypeProperty,
+          source: None,
+        })?;
+      }
+    };
+
+    match credential.expiration_date {
+      None => {
+        return Err(DomainLinkageVerificationError {
+          cause: DomainLinkageVerificationErrorCause::MissingExpirationDate,
+          source: None,
+        });
+      }
+      Some(expiration_date) => {
+        CredentialValidator::check_expires_on_or_after(credential, expiration_date).map_err(|err| {
+          DomainLinkageVerificationError {
+            cause: DomainLinkageVerificationErrorCause::CredentialValidationError,
+            source: Some(Box::new(err)),
+          }
+        })?;
+      }
+    }
+
+    match &credential.credential_subject {
+      OneOrMany::One(credential_subject) => {
+        match &credential_subject.id {
+          None => {
+            return Err(DomainLinkageVerificationError {
+              cause: DomainLinkageVerificationErrorCause::MissingSubjectId,
+              source: None,
+            });
+          }
+          Some(id) => match CoreDID::parse(id.to_string()) {
+            Ok(subject_did) => {
+              if issuer_did != subject_did {
+                return Err(DomainLinkageVerificationError {
+                  cause: DomainLinkageVerificationErrorCause::IssuerSubjectMismatch,
+                  source: None,
+                });
+              }
+            }
+            Err(_) => {
+              return Err(DomainLinkageVerificationError {
+                cause: DomainLinkageVerificationErrorCause::InvalidSubjectId,
+                source: None,
+              });
+            }
+          },
+        }
+
+        let origin: &str = credential_subject
+          .properties
+          .get("origin")
+          .ok_or(DomainLinkageVerificationError {
+            cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
+            source: None,
+          })?
+          .as_str()
+          .ok_or(DomainLinkageVerificationError {
+            cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
+            source: None,
+          })?;
+        let origin_url = Url::parse(origin)
+          .map_err(|_err| Url::parse("https://".to_owned() + origin))
+          .map_err(|_err| DomainLinkageVerificationError {
+            cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
+            source: None,
+          })?;
```

Comment:
This maps the error of the first `Url::parse` to the result of the second `Url::parse` which is not what we want. It also always retries to parse the Url no matter what error is returned.
Suggestion:

```suggestion
     let origin_url: Url = match Url::parse(origin) {
          Ok(url) => url,
          Err(identity_core::Error::InvalidUrl(url::ParseError::RelativeUrlWithoutBase)) => {
            match Url::parse("https://".to_owned() + origin) {
              Ok(url) => url,
              Err(err) => {
                return Err(DomainLinkageVerificationError {
                  cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
                  source: Some(Box::new(err)),
                });
              }
            }
          }
          Err(err) => {
            return Err(DomainLinkageVerificationError {
              cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
              source: Some(Box::new(err)),
            });
          }
        };
```

Requires `url = { version = "2.2", default-features = false }` in the `Cargo.toml`.